ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs

```diff
@@ -36,28 +41,64 @@ import           Ouroboros.Consensus.Ledger.Byron.Config
 
 import qualified Test.Cardano.Chain.Genesis.Dummy as Dummy
 
+{-------------------------------------------------------------------------------
+  Conditions
+-------------------------------------------------------------------------------}
+
+data ByronProtocolSetupError =
+      ByronInconsistentLeaderCredentials PbftLeaderCredentials
+      deriving (Eq, Show)
+
+instance Exception ByronProtocolSetupError
+
 {-------------------------------------------------------------------------------
   Credentials
 -------------------------------------------------------------------------------}
 
 data PbftLeaderCredentials = PbftLeaderCredentials {
       plcSignKey     :: Cardano.SigningKey
     , plcDlgCert     :: Cardano.Delegation.Certificate
-    }
+    } deriving (Eq, Show)
 
 {-------------------------------------------------------------------------------
   ProtocolInfo
 -------------------------------------------------------------------------------}
 
+-- | Given a genesis delegation and a delegate VK, try to find its delegation certificate.
+--   Note that in Byron era, the delegation map is injective, so there's at most once such
+--   certificate.
+--   TODO:  Move to cardano-ledger.
+findCertByDelegateByron :: Cardano.Genesis.GenesisDelegation -> Cardano.VerificationKey -> Maybe Cardano.Delegation.Certificate
+findCertByDelegateByron Cardano.Genesis.UnsafeGenesisDelegation{Cardano.Genesis.unGenesisDelegation=dlg} =
+  flip List.find (Map.elems dlg) . certDelegatesToVK
+  where
+    certDelegatesToVK :: Cardano.VerificationKey -> Cardano.Delegation.Certificate -> Bool
+    certDelegatesToVK vk = (== vk) . Cardano.Delegation.delegateVK
+
+-- TODO:  What happens to node's PBFT id when the delegation map changes at runtime?
+-- | Given a delegation certificate, assign it a PBFT id, in the context of a fixed delegation map.
+--   The order is defined by the delegator's KeyHash ordering.
+certificateNodeIdPbft :: Cardano.Genesis.GenesisDelegation -> Cardano.Delegation.Certificate -> Maybe CoreNodeId
+certificateNodeIdPbft Cardano.Genesis.UnsafeGenesisDelegation{Cardano.Genesis.unGenesisDelegation=dlg} cert =
+  CoreNodeId <$> List.findIndex ((== cert) . snd) (Map.toAscList dlg)
+
+-- | Given a delegate key, assign it a PBFT id, in the context of a fixed delegation map.
+--   The order is defined by the delegator's KeyHash ordering.
+delegateNodeIdPbft :: Cardano.Genesis.GenesisDelegation -> Cardano.VerificationKey -> Maybe CoreNodeId
+delegateNodeIdPbft dlg vk = join $ certificateNodeIdPbft dlg <$> findCertByDelegateByron dlg vk
```

Comment:
I'm pretty sure this is the wrong algorithm. The right algorithm was in my original WIP patch.

What we should do is this:
* take the supplied delegation cert, grab the genesis key (hash) from that
* look up that genesis key (hash) in the set of known genesis key hashes (from the genesis file) and if present, determine the index.

```
+    genesisKeyHashes :: Set Cardano.Common.KeyHash
+    genesisKeyHashes = Cardano.Genesis.unGenesisKeyHashes
+                     . Cardano.Genesis.configGenesisKeyHashes
+                     $ genesisConfig
+
+    -- The CoreNodeId index is determined by the 0-based position in the
+    -- sort order of the genesis key hashes.
+    -- TODO: this is important policy, it should live in the core PBFT module.
+    genesisKeyCoreNodeId :: Cardano.ProtocolMagicId
+                         -> VerKeyDSIGN CardanoDSIGN
+                         -> Maybe CoreNodeId
+    genesisKeyCoreNodeId magic vkey =
+        Data.Reflection.give magic $
+        CoreNodeId <$> Set.lookupIndex (hashVerKey vkey) genesisKeyHashes
```

Reply:
Done.